iceoryx_posh/include/iceoryx_posh/popo/sample.hpp

```diff
@@ -27,123 +28,155 @@ namespace popo
-    /// @details Only available for non void type T.
+    /// @details Only available for non-const type T.
     ///
-    template <typename S = T,
-              typename = std::enable_if_t<std::is_same<S, T>::value && !std::is_same<S, void>::value, S>>
+    template <typename S = T, typename = ForPublisherOnly<S, T>>
     S& operator*() noexcept;
 
     ///
     /// @brief operator* Provide a const reference to the encapsulated type.
-    /// @return A T& to the encapsulated type.
-    /// @details Only available for non void type T.
+    /// @return A const T& to the encapsulated type.
     ///
-    template <typename S = T,
-              typename = std::enable_if_t<std::is_same<S, T>::value && !std::is_same<S, void>::value, S>>
-    const S& operator*() const noexcept;
+    const T& operator*() const noexcept;
 
     ///
     /// @brief operator bool Indciates whether the sample is valid, i.e. refers to allocated memory.
     /// @return true if the sample is valid, false otherwise.
     ///
-    operator bool() const;
+    operator bool() const noexcept;
 
     ///
-    /// @brief allocation Access to the memory chunk loaned to the sample.
-    /// @return
+    /// @brief allocation Access to the encapsulated type loaned to the sample.
+    /// @return a pointer to the encapsulated type.
+    /// @details Only available for non-const type T.
     ///
+    template <typename S = T, typename = ForPublisherOnly<S, T>>
     T* get() noexcept;
 
     ///
-    /// @brief allocation Read-only access to the memory chunk loaned to the sample.
-    /// @return
+    /// @brief allocation Read-only access to the encapsulated type loaned to the sample.
+    /// @return a const pointer to the encapsulated type.
     ///
     const T* get() const noexcept;
 
     ///
     /// @brief header Retrieve the header of the underlying memory chunk loaned to the sample.
     /// @return The ChunkHeader of the underlying memory chunk.
+    /// @details Only available for non-const type T.
     ///
+    template <typename S = T, typename = ForPublisherOnly<S, T>>
     mepoo::ChunkHeader* getHeader() noexcept;
 
+    ///
+    /// @brief header Retrieve the header of the underlying memory chunk loaned to the sample.
+    /// @return The const ChunkHeader of the underlying memory chunk.
+    ///
+    const mepoo::ChunkHeader* getHeader() const noexcept;
+
     ///
     /// @brief publish Publish the sample via the publisher from which it was loaned and automatically
     /// release ownership to it.
+    /// @details Only available for non-const type T.
     ///
+    template <typename S = T, typename = ForPublisherOnly<S, T>>
     void publish() noexcept;
 
     ///
-    /// @brief release Manually release ownership of the loaned memory chunk.
-    /// @details This prevents the sample from automatically releasing ownership on destruction.
+    /// @brief release Manually transfers ownership of the loaned memory chunk back to the mempool.
```

Comment:
This comment is not quite right. Calling `releaseChunk()` would transfer the chunk back to the memory pool, which this method does not do.

Reply:
That's exactly the reason I assumed this was wrong. It's the `release` semantic of the `unique_ptr` not of the `chunk`

Reply:
I slept over the used term here and I would recommend to follow the STL semantic because this is the common semantic in the C++ world. This means that `release` should be named `reset` since every C++ developer would expect that it behaves like that and the sample is somehow also a unique ptr.
I know that we also use `releaseChunk` but in this case I would break the semantics here since the name is also somehow different and it is not the name of another unique ptr like structure.

Reply:
@elfenpiff I got it wrong. The `Sample::release` method should do exactly what `unique_ptr::release` does, releasing the ownership to the caller. After moving the untyped API away from `Sample<void>` and having the `Sample` only for the typed API, this should probably not be accessible by the user anymore, just by the publisher.

Reply:
I don't get it. This `release()` does not return the resource like a unique_ptr. And @elBoberido changed the internal to do a reset of the unique_ptr. I think because a releae on the unique_ptr without taking care of the returned resource is a memory leak in our case. 
So from what it is doing now it is a reset. But why do we need this? If `publish()` is not called, the Sample d'tor will return the chunk to the mempool without publishing it. Is it an early release? But what is a `Sample` good for that has no chunk?

Reply:
@budrus to clarify this a bit. I wrongly assumed that he `Sample::release` has the same semantic as `releaseChunk` and therefore called `unique_ptr::reset` instead of `unique_ptr::release` but this is wrong since the publisher needs to gain ownership of the chunk from the `Sample` and therefore needs to call `Sample::release` without invoking the custom deleter. Previously the `Sample::release` was also necessary for the untyped API, but this is not the case anymore. To make a long story short, `Sample::release` should be private and only be accessible by the `Publisher`, else this will lead to chunk leaks.

If someone want's to control the lifetime of the `Sample` manually, a container must be used, e.g. `cxx::optional<Sample<T>>`.

Reply:
@elBoberido Got it. Then you found a way to create a leak and the solution is that only the publisher is allowed to call `release` which we should maybe rename to `reset`

Reply:
@elBoberido You changed it to release the unique_ptr again. But the `Sample` `release` does not provide the resource. 
Should we 
* call reset and name the method in Sample also reset?
* call it release and return the resource and the publisher than calls release to get the ownership of the chunk?

Reply:
@budrus currently the publisher uses `Sample::get` to get a raw pointer to the chunk. But yes, it might be clearer what happens if `Sample::release` returns the raw pointer. Then the API would also be more similar to `unique_ptr::release`